{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "zsh:1: command not found: pip\r\n"
     ]
    }
   ],
   "source": [
    "!pip install myvariant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'myvariant'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-f8a6bef61f3d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# imports\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mmyvariant\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'myvariant'"
     ]
    }
   ],
   "source": [
    "\n",
    "# imports\n",
    "import myvariant\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "#from jupyter_datatables import init_datatables_mode\n",
    "\n",
    "#init_datatables_mode()\n",
    "mv = myvariant.MyVariantInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-7fce8d450a54>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mfeather_file_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"data/sm_feather\"\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_feather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeather_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muse_threads\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m## Narrow it down\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchromosome\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"6\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "feather_file_path = \"data/sm_feather\";\n",
    "df = pd.read_feather(feather_file_path, use_threads=True);\n",
    "\n",
    "## Narrow it down\n",
    "df = df[df.chromosome == \"6\"]\n",
    "\n",
    "# GeneLoc map region around 10MB CYP21 : 27038265 --> 37041670\n",
    "#df = df[(df['position'] >= 27038265) & (df['position'] <= 37041670)] # 7000 entries\n",
    "#df = df[(df['position'] >= 33038265) & (df['position'] <= 35041670)] # 600 entries\n",
    "\n",
    "# Remove unneeeded columns\n",
    "#del df['position']\n",
    "#del df['chromosome']\n",
    "\n",
    "# df_Full.to_feather(feather_file_path);\n",
    "df.reset_index(drop=True)\n",
    "\n",
    "# Remove i-rsids\n",
    "df.drop(df[df.rsid.str.contains('i')].index, inplace=True)\n",
    "\n",
    "# Move to a list for mv.querymany \n",
    "your_rsid_list = df['rsid'].values.tolist()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import from CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene</th>\n",
       "      <th>alt</th>\n",
       "      <th>adviser_score</th>\n",
       "      <th>_id</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6:g.137236687G&gt;A</td>\n",
       "      <td>5~PEX7~Uncommon, Predicted Neutral</td>\n",
       "      <td>A</td>\n",
       "      <td>PEX7</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0175</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9825</td>\n",
       "      <td>rs72987544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6:g.135899344C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C6orf217</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0825</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9175</td>\n",
       "      <td>rs2092556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6:g.149749403G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>MAP3K7IP2</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4175</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5825</td>\n",
       "      <td>rs392917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6:g.28485680A&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>GPX6</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9950</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0050</td>\n",
       "      <td>rs423118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6:g.98980147A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>POU3F2</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1850</td>\n",
       "      <td>G</td>\n",
       "      <td>0.8150</td>\n",
       "      <td>rs211222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37707</th>\n",
       "      <td>chr6:g.132879437A&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>TAAR8</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9750</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0250</td>\n",
       "      <td>rs77325751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37708</th>\n",
       "      <td>chr6:g.474405C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>EXOC2</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1800</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8200</td>\n",
       "      <td>rs11242905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37709</th>\n",
       "      <td>chr6:g.109169044C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>ARMC2</td>\n",
       "      <td>C</td>\n",
       "      <td>0.8225</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1775</td>\n",
       "      <td>rs948522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37710</th>\n",
       "      <td>chr6:g.90021186T&gt;C</td>\n",
       "      <td>5~GABRR2~Uncommon, Predicted Neutral</td>\n",
       "      <td>C</td>\n",
       "      <td>GABRR2</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0375</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9625</td>\n",
       "      <td>rs79482608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37711</th>\n",
       "      <td>chr6:g.74148730C&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>C6orf150</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9725</td>\n",
       "      <td>rs7763873</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37712 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      rsid                                  gene alt  \\\n",
       "0      chr6:g.137236687G>A    5~PEX7~Uncommon, Predicted Neutral   A   \n",
       "1      chr6:g.135899344C>T                                   NaN   T   \n",
       "2      chr6:g.149749403G>A                                   NaN   A   \n",
       "3       chr6:g.28485680A>C                                   NaN   C   \n",
       "4       chr6:g.98980147A>G                                   NaN   G   \n",
       "...                    ...                                   ...  ..   \n",
       "37707  chr6:g.132879437A>C                                   NaN   C   \n",
       "37708     chr6:g.474405C>T                                   NaN   T   \n",
       "37709  chr6:g.109169044C>T                                   NaN   T   \n",
       "37710   chr6:g.90021186T>C  5~GABRR2~Uncommon, Predicted Neutral   C   \n",
       "37711   chr6:g.74148730C>A                                   NaN   A   \n",
       "\n",
       "      adviser_score _id  Allele1 Freq1  Allele2       Freq2  \n",
       "0              PEX7   A   0.0175     G   0.9825  rs72987544  \n",
       "1          C6orf217   C   0.0825     T   0.9175   rs2092556  \n",
       "2         MAP3K7IP2   A   0.4175     G   0.5825    rs392917  \n",
       "3              GPX6   A   0.9950     C   0.0050    rs423118  \n",
       "4            POU3F2   A   0.1850     G   0.8150    rs211222  \n",
       "...             ...  ..      ...   ...      ...         ...  \n",
       "37707         TAAR8   A   0.9750     C   0.0250  rs77325751  \n",
       "37708         EXOC2   C   0.1800     T   0.8200  rs11242905  \n",
       "37709         ARMC2   C   0.8225     T   0.1775    rs948522  \n",
       "37710        GABRR2   C   0.0375     T   0.9625  rs79482608  \n",
       "37711      C6orf150   A   0.0275     C   0.9725   rs7763873  \n",
       "\n",
       "[37712 rows x 9 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = pd.read_csv(\"data/23andme_v5_ch6.csv\") \n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene</th>\n",
       "      <th>alt</th>\n",
       "      <th>adviser_score</th>\n",
       "      <th>_id</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6:g.137236687G&gt;A</td>\n",
       "      <td>5~PEX7~Uncommon, Predicted Neutral</td>\n",
       "      <td>A</td>\n",
       "      <td>PEX7</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0175</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9825</td>\n",
       "      <td>rs72987544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6:g.135899344C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C6orf217</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0825</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9175</td>\n",
       "      <td>rs2092556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6:g.149749403G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>MAP3K7IP2</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4175</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5825</td>\n",
       "      <td>rs392917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6:g.28485680A&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>GPX6</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9950</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0050</td>\n",
       "      <td>rs423118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6:g.98980147A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>POU3F2</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1850</td>\n",
       "      <td>G</td>\n",
       "      <td>0.8150</td>\n",
       "      <td>rs211222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37707</th>\n",
       "      <td>chr6:g.132879437A&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>TAAR8</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9750</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0250</td>\n",
       "      <td>rs77325751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37708</th>\n",
       "      <td>chr6:g.474405C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>EXOC2</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1800</td>\n",
       "      <td>T</td>\n",
       "      <td>0.8200</td>\n",
       "      <td>rs11242905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37709</th>\n",
       "      <td>chr6:g.109169044C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>ARMC2</td>\n",
       "      <td>C</td>\n",
       "      <td>0.8225</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1775</td>\n",
       "      <td>rs948522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37710</th>\n",
       "      <td>chr6:g.90021186T&gt;C</td>\n",
       "      <td>5~GABRR2~Uncommon, Predicted Neutral</td>\n",
       "      <td>C</td>\n",
       "      <td>GABRR2</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0375</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9625</td>\n",
       "      <td>rs79482608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37711</th>\n",
       "      <td>chr6:g.74148730C&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>C6orf150</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9725</td>\n",
       "      <td>rs7763873</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37712 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      rsid                                  gene alt  \\\n",
       "0      chr6:g.137236687G>A    5~PEX7~Uncommon, Predicted Neutral   A   \n",
       "1      chr6:g.135899344C>T                                   NaN   T   \n",
       "2      chr6:g.149749403G>A                                   NaN   A   \n",
       "3       chr6:g.28485680A>C                                   NaN   C   \n",
       "4       chr6:g.98980147A>G                                   NaN   G   \n",
       "...                    ...                                   ...  ..   \n",
       "37707  chr6:g.132879437A>C                                   NaN   C   \n",
       "37708     chr6:g.474405C>T                                   NaN   T   \n",
       "37709  chr6:g.109169044C>T                                   NaN   T   \n",
       "37710   chr6:g.90021186T>C  5~GABRR2~Uncommon, Predicted Neutral   C   \n",
       "37711   chr6:g.74148730C>A                                   NaN   A   \n",
       "\n",
       "      adviser_score _id  Allele1 Freq1  Allele2       Freq2  \n",
       "0              PEX7   A   0.0175     G   0.9825  rs72987544  \n",
       "1          C6orf217   C   0.0825     T   0.9175   rs2092556  \n",
       "2         MAP3K7IP2   A   0.4175     G   0.5825    rs392917  \n",
       "3              GPX6   A   0.9950     C   0.0050    rs423118  \n",
       "4            POU3F2   A   0.1850     G   0.8150    rs211222  \n",
       "...             ...  ..      ...   ...      ...         ...  \n",
       "37707         TAAR8   A   0.9750     C   0.0250  rs77325751  \n",
       "37708         EXOC2   C   0.1800     T   0.8200  rs11242905  \n",
       "37709         ARMC2   C   0.8225     T   0.1775    rs948522  \n",
       "37710        GABRR2   C   0.0375     T   0.9625  rs79482608  \n",
       "37711      C6orf150   A   0.0275     C   0.9725   rs7763873  \n",
       "\n",
       "[37712 rows x 9 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['rsid', 'gene', 'alt', 'adviser_score', '_id', 'Allele1', 'Freq1', 'Allele2', 'Freq2', 'chromosome_x', 'position_x', 'genotype_x', 'chromosome_y', 'position_y', 'genotype_y', 'chromosome', 'position', 'genotype']\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "'wellderly.gene' is not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-384e68d1c0fe>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'rsid'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'wellderly.gene'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'genotype'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'wellderly.alt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: 'wellderly.gene' is not in list"
     ]
    }
   ],
   "source": [
    "# Merge your data 'rsid'\n",
    "res = res.merge(df, on='rsid', how='outer')\n",
    "cols = list(res.columns.values)\n",
    "print(cols)\n",
    "\n",
    "# Move rsid/genotype to front\n",
    "\n",
    "#cols = res.columns.tolist()\n",
    "#cols\n",
    "#cols.insert(1, cols.pop(cols.index('genotype')))\n",
    "#res = res.reindex(columns= cols)\n",
    "\n",
    "\n",
    "# Move rsid/genotype to front\n",
    "\n",
    "cols = res.columns.tolist()\n",
    "cols\n",
    "cols.insert(0, cols.pop(cols.index('rsid')))\n",
    "cols.insert(1, cols.pop(cols.index('wellderly.gene')))\n",
    "cols.insert(2, cols.pop(cols.index('genotype')))\n",
    "cols.insert(3, cols.pop(cols.index('wellderly.alt')))\n",
    "cols.insert(4, cols.pop(cols.index('wellderly.ref')))\n",
    "cols.insert(5, cols.pop(cols.index('wellderly.adviser_score')))\n",
    "res = res.reindex(columns= cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res.columns =  ['rsid', 'gene', 'genotype', 'alt', 'ref',  'adviser_score', '_id', 'alleles']\n",
    "\n",
    "# Fill NaN with '--'\n",
    "res_merged = res_merged.fillna(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Match!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill NaN \n",
    "res = res.fillna(\"\")\n",
    "\n",
    "# Match\n",
    "res = res[res.apply(lambda x: x.alt in x.genotype, axis=1)]\n",
    "\n",
    "res = res.reset_index(drop=True)\n",
    "\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_html = res\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HTML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean\n",
    "res_html['adviser_score'] = res_html['adviser_score'].str.replace(r\"\\[\",\"\")\n",
    "res_html['adviser_score'] = res_html['adviser_score'].str.replace(r\"\\]\",\"\")\n",
    "\n",
    "# Fill NA\n",
    "\n",
    "res_html = res_html.fillna(' ')\n",
    "\n",
    "# Sort\n",
    "res_html = res_html.sort_values('adviser_score', ascending=False)\n",
    "\n",
    "# to HTML\n",
    "html = res_html.to_html()\n",
    "\n",
    "# Write HTML \n",
    "text_file = open(\"susan_ch6.html\", \"w\")\n",
    "text_file.write(html)\n",
    "text_file.close()\n",
    "\n",
    "res_html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(\"data/10mb_cyp21.csv\") \n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save to CSV\n",
    "\n",
    "res.to_csv('data/10mb_cyp21.csv', encoding='utf-8', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
