{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA Panda\n",
    "\n",
    "This is a small script that will import your genome, and query specified genes against NCBI returning a data_frame and .csv with positive matches. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import pandas as pd\n",
    "from os import listdir\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import pylev\n",
    "\n",
    "import re\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_style('darkgrid')\n",
    "sns.color_palette('Spectral')\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "import requests\n",
    "\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.support.ui import WebDriverWait"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import User Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_frame = []\n",
    "\n",
    "user_frame.append(pd.read_csv('data/23andme_MG_v4.txt', sep='\\t', \n",
    "                   dtype={'rsid':'str', 'chromosome':'object', 'position':'int', 'genotype':'str'}, \n",
    "                   comment='#'))\n",
    "\n",
    "\n",
    "data_frame = pd.concat(user_frame, axis=0, ignore_index=True)\n",
    "\n",
    "\n",
    "#import_frame = pd.read_csv(\"rccx.csv\") \n",
    "#merged_frame = pd.concat([data_frame, import_frame], axis=0, sort=True)\n",
    "#print(merged_frame)\n",
    "#df = pd.DataFrame(merged_frame)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method DataFrame.info of                rsid chromosome  position genotype\n",
       "0        rs12564807          1    734462       GG\n",
       "1         rs3131972          1    752721       GG\n",
       "2       rs148828841          1    760998       CC\n",
       "3        rs12124819          1    776546       AA\n",
       "4       rs115093905          1    787173       GT\n",
       "...             ...        ...       ...      ...\n",
       "601866     i4000757         MT     16526        G\n",
       "601867      i701671         MT     16526        G\n",
       "601868     i4990307         MT     16527        C\n",
       "601869     i4000756         MT     16540        C\n",
       "601870     i3001931         MT     16547        C\n",
       "\n",
       "[601871 rows x 4 columns]>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the data into a pandas DataFrame and do some EDA\n",
    "df = pd.DataFrame(data_frame)\n",
    "#df = pd.DataFrame(merged_frame)\n",
    "df.info\n",
    "#df = df.fillna(\"0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rsid          601871\n",
       "chromosome        25\n",
       "position      597380\n",
       "genotype          20\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df.isna().any()\n",
    "# How many chromosomes are on the Y chromosome?\n",
    "df['chromosome'].unique()\n",
    "Y_chromosome = df[df.chromosome == 'Y']\n",
    "len(Y_chromosome)\n",
    "# Show unique counts\n",
    "df.nunique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Display how many missing SNPs are in your genome\n",
    "genotype_na = df[df.genotype == '--']\n",
    "len(genotype_na)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 40384 entries, 199607 to 239990\n",
      "Data columns (total 4 columns):\n",
      "rsid          40384 non-null object\n",
      "chromosome    40384 non-null object\n",
      "position      40384 non-null int64\n",
      "genotype      40384 non-null object\n",
      "dtypes: int64(1), object(3)\n",
      "memory usage: 1.5+ MB\n"
     ]
    }
   ],
   "source": [
    "# Print the length of any chromosome\n",
    "df6 = df[df.chromosome == \"6\"]\n",
    "len(df6)\n",
    "df6.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>199607</th>\n",
       "      <td>rs12209455</td>\n",
       "      <td>6</td>\n",
       "      <td>167510</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199608</th>\n",
       "      <td>rs6596796</td>\n",
       "      <td>6</td>\n",
       "      <td>175522</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199609</th>\n",
       "      <td>rs1535053</td>\n",
       "      <td>6</td>\n",
       "      <td>181089</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199610</th>\n",
       "      <td>rs6937355</td>\n",
       "      <td>6</td>\n",
       "      <td>183900</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199611</th>\n",
       "      <td>rs6937363</td>\n",
       "      <td>6</td>\n",
       "      <td>183917</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              rsid chromosome  position genotype\n",
       "199607  rs12209455          6    167510       AG\n",
       "199608   rs6596796          6    175522       GG\n",
       "199609   rs1535053          6    181089       TT\n",
       "199610   rs6937355          6    183900       CC\n",
       "199611   rs6937363          6    183917       CC"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df6.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 39 entries, 210570 to 203706\n",
      "Data columns (total 4 columns):\n",
      "chromosome    39 non-null object\n",
      "genotype      39 non-null object\n",
      "position      39 non-null int64\n",
      "rsid          39 non-null object\n",
      "dtypes: int64(1), object(3)\n",
      "memory usage: 1.5+ KB\n"
     ]
    }
   ],
   "source": [
    "# See the frequency of genotypes\n",
    "#df6['genotype'].value_counts()\n",
    "df6.count()\n",
    "notch4 = df6[(df6['position'] >= 32194843) & (df6['position'] <= 32224067)]\n",
    "notch3 = df[(df['position'] >= 15159038) & (df6['position'] <= 15200995)]\n",
    "notch4.count()\n",
    "notch3.count()\n",
    "notch = pd.concat([notch4, notch3], axis=0, sort=True)\n",
    "notch.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Isolate the RCCX module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CYP21A2 :: 32,038,306 to 32,041,670 on chromosome 6 \n",
    "\n",
    "# tnxb :: 32,041,153 to 32,109,338 \n",
    "\n",
    "#Â C4 :: 31,982,057 to 32,002,681\n",
    "\n",
    "# stk19 :: 31,971,175 to 31,981,446 \n",
    "\n",
    "# NOTCH 3 :: 15,159,038 to 15,200,995\n",
    "\n",
    "# NOTCH 4 32,194,843 to 32,224,067"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rsid          75\n",
       "chromosome    75\n",
       "position      75\n",
       "genotype      75\n",
       "dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rccx = df6[(df6['position'] >= 31971175) & (df6['position'] <= 32109338)]\n",
    "rccx = rccx[rccx.genotype != \"--\"]\n",
    "rccx.count()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GG    40\n",
       "CC    33\n",
       "TT    13\n",
       "AA     9\n",
       "CT     9\n",
       "AG     5\n",
       "CG     3\n",
       "AT     2\n",
       "Name: genotype, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "toScan = pd.concat([notch, rccx], axis=0, sort=True)\n",
    "\n",
    "toScan['genotype'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chromosome    114\n",
       "genotype      114\n",
       "position      114\n",
       "rsid          114\n",
       "dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.options.display.max_rows = 999\n",
    "toScan.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Crawling NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trying... rs3096691 ( 1 out of 39 )\n",
      "Risk G\n",
      "Frequency .459\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs482759 ( 2 out of 39 )\n",
      "Risk G\n",
      "Frequency .187\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs365053 ( 3 out of 39 )\n",
      "Risk G\n",
      "Frequency .252\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs495089 ( 4 out of 39 )\n",
      "Risk G\n",
      "Frequency .340\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs436845 ( 5 out of 39 )\n",
      "Risk G\n",
      "Frequency .277\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs404890 ( 6 out of 39 )\n",
      "Risk A\n",
      "Frequency .334\n",
      "Gene None\n",
      "Publications 1\n",
      "trying... rs2849015 ( 7 out of 39 )\n",
      "Risk A\n",
      "Frequency .377\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs9267873 ( 8 out of 39 )\n",
      "Risk C\n",
      "Frequency .193\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs3134926 ( 9 out of 39 )\n",
      "Risk G\n",
      "Frequency .353\n",
      "Gene None\n",
      "Publications 1\n",
      "trying... rs12182351 ( 10 out of 39 )\n",
      "index error\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs3130299 ( 11 out of 39 )\n",
      "Risk G\n",
      "Frequency .240\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs549182 ( 12 out of 39 )\n",
      "Risk A\n",
      "Frequency .080\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs549270 ( 13 out of 39 )\n",
      "Risk T\n",
      "Frequency .109\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs9267911 ( 14 out of 39 )\n",
      "Risk C\n",
      "Frequency .447\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs6901158 ( 15 out of 39 )\n",
      "Risk T\n",
      "Frequency .094\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs3130304 ( 16 out of 39 )\n",
      "Risk A\n",
      "Frequency .199\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs416352 ( 17 out of 39 )\n",
      "Risk T\n",
      "Frequency .378\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs424232 ( 18 out of 39 )\n",
      "Risk T\n",
      "Frequency .320\n",
      "Gene None\n",
      "Publications 2\n",
      "trying... rs382259 ( 19 out of 39 )\n",
      "https://www.ncbi.nlm.nih.gov/snp/rs382259#clinical_significance was not found or on dbSNP or contained no valid information\n",
      "trying... rs507778 ( 20 out of 39 )\n",
      "Risk T\n",
      "Frequency .432\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs412657 ( 21 out of 39 )\n",
      "Risk T\n",
      "Frequency .367\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs9267947 ( 22 out of 39 )\n",
      "Risk A\n",
      "Frequency .376\n",
      "Gene None\n",
      "Publications 1\n",
      "trying... rs411326 ( 23 out of 39 )\n",
      "Risk T\n",
      "Frequency .290\n",
      "Gene None\n",
      "Publications 2\n",
      "trying... rs17576984 ( 24 out of 39 )\n",
      "Risk T\n",
      "Frequency .118\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs440169 ( 25 out of 39 )\n",
      "Risk A\n",
      "Frequency .447\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs1559876 ( 26 out of 39 )\n",
      "Risk C\n",
      "Frequency .323\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs2894239 ( 27 out of 39 )\n",
      "Risk G\n",
      "Frequency .356\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs2395110 ( 28 out of 39 )\n",
      "Risk A\n",
      "Frequency .474\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs6936204 ( 29 out of 39 )\n",
      "Risk T\n",
      "Frequency .237\n",
      "Gene None\n",
      "Publications 2\n",
      "trying... rs3115573 ( 30 out of 39 )\n",
      "Risk G\n",
      "Frequency .447\n",
      "Gene None\n",
      "Publications 1\n",
      "trying... rs9296015 ( 31 out of 39 )\n",
      "Risk A\n",
      "Frequency .245\n",
      "Gene None\n",
      "Publications 1\n",
      "trying... rs4959089 ( 32 out of 39 )\n",
      "Risk G\n",
      "Frequency .245\n",
      "Gene None\n",
      "Publications 0\n",
      "trying... rs3115572 ( 33 out of 39 )\n",
      "Risk C\n",
      "Frequency .397\n",
      "Gene None\n",
      "Publications 1\n",
      "trying... rs3130315 ( 34 out of 39 )\n",
      "Risk A\n",
      "Frequency .399\n",
      "Gene None\n",
      "Publications 3\n",
      "trying... rs3130320 ( 35 out of 39 )\n",
      "Risk T\n",
      "Frequency .271\n",
      "Gene LOC101929163\n",
      "Publications 3\n",
      "trying... rs9367866 ( 36 out of 39 )\n",
      "Risk T\n",
      "Frequency .081\n",
      "Gene LOC105374946\n",
      "Publications 0\n",
      "trying... rs6935115 ( 37 out of 39 )\n",
      "Risk C\n",
      "Frequency .184\n",
      "Gene LOC105374946\n",
      "Publications 0\n",
      "trying... rs79674673 ( 38 out of 39 )\n",
      "Risk T\n",
      "Frequency .058\n",
      "Gene LOC105374946\n",
      "Publications 0\n",
      "trying... rs7747637 ( 39 out of 39 )\n",
      "Risk C\n",
      "Frequency .081\n",
      "Gene LOC105374946\n",
      "Publications 0\n"
     ]
    }
   ],
   "source": [
    "import urllib.request\n",
    "from bs4 import BeautifulSoup\n",
    "count = 0\n",
    "toScan['Parsed'] = \"0\"\n",
    "\n",
    "for i, row in toScan.iterrows():\n",
    "    count = count + 1\n",
    "    if(row.Parsed != \"1\"):\n",
    "        try:\n",
    "            print(\"trying...\", row.rsid,\"(\", count, \"out of\", len(rccx['rsid']),\")\")\n",
    "            url = \"https://www.ncbi.nlm.nih.gov/snp/\" + row.rsid + \"#clinical_significance\"\n",
    "            response = urllib.request.urlopen(url)\n",
    "            html = response.read()\n",
    "            bs = BeautifulSoup(html, \"html.parser\")\n",
    "\n",
    "            classification = bs.find(id=\"clinical_significance\")\n",
    "\n",
    "            if classification:\n",
    "                rows = classification.find_all(\"tr\")\n",
    "                ClinVar = []\n",
    "                for row in rows:\n",
    "                    cols = row.find_all(\"td\")\n",
    "                    cols = [ele.text.strip() for ele in cols]\n",
    "                    ClinVar.append([ele for ele in cols if ele])\n",
    "                listToStr = ' '.join([str(elem) for elem in ClinVar]) \n",
    "\n",
    "                toScan.at[i, 'ClinVar'] = listToStr\n",
    "\n",
    "            ncbi = bs.find(class_=\"summary-box usa-grid-full\")\n",
    "            if ncbi:\n",
    "                dbSNP = []\n",
    "\n",
    "                rows = ncbi.find_all(\"div\")\n",
    "\n",
    "                for row in rows:\n",
    "                    cols = row.find_all(\"div\")\n",
    "                    cols = [ele.text.strip() for ele in cols]\n",
    "                    dbSNP.append(cols)\n",
    "\n",
    "                try:\n",
    "                    print(\"Risk\", dbSNP[2][0][0])\n",
    "                    print(\"Frequency\",dbSNP[2][0][3:7])\n",
    "\n",
    "                    toScan.at[i, 'Risk'] = dbSNP[2][0][0]\n",
    "                    toScan.at[i, 'Frequency'] = dbSNP[2][0][3:7]\n",
    "                except IndexError:\n",
    "                    print(\"index error\")\n",
    "\n",
    "                dbSNPTwo= []\n",
    "                rows = ncbi.find_all(\"dl\")\n",
    "\n",
    "                for row in rows:\n",
    "                    cols = row.find_all(\"dd\")\n",
    "                    cols = [ele.text.strip() for ele in cols]\n",
    "                    dbSNPTwo.append(cols)\n",
    "\n",
    "            try:\n",
    "                print(\"Gene\", dbSNPTwo[1][1].split(' ')[0])\n",
    "                toScan.at[i, 'Gene'] = dbSNPTwo[1][1].split(' ')[0]\n",
    "                print(\"Publications\", dbSNPTwo[1][2][0]) \n",
    "                toScan.at[i, 'Citations'] = dbSNPTwo[1][2][0]\n",
    "                toScan.at[i, 'Parsed'] = \"1\"\n",
    "            except IndexError:\n",
    "                    print(\"index error\")\n",
    "\n",
    "\n",
    "        except urllib.error.HTTPError:\n",
    "            print(url + \" was not found or on dbSNP or contained no valid information\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rccx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rccx.to_csv('rccx.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rccx_filled = rccx.fillna(\"0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rccx_filled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rccx_present = rccx_filled\n",
    "rccx_present = rccx_filled[rccx_filled.apply(lambda x: x.Risk in x.genotype, axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rccx_present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genotype</th>\n",
       "      <th>position</th>\n",
       "      <th>rsid</th>\n",
       "      <th>ClinVar</th>\n",
       "      <th>Risk</th>\n",
       "      <th>Frequency</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Citations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>210570</th>\n",
       "      <td>6</td>\n",
       "      <td>AG</td>\n",
       "      <td>32194854</td>\n",
       "      <td>rs3096691</td>\n",
       "      <td></td>\n",
       "      <td>G</td>\n",
       "      <td>.459</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210573</th>\n",
       "      <td>6</td>\n",
       "      <td>AG</td>\n",
       "      <td>32197463</td>\n",
       "      <td>rs495089</td>\n",
       "      <td></td>\n",
       "      <td>G</td>\n",
       "      <td>.340</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210574</th>\n",
       "      <td>6</td>\n",
       "      <td>AG</td>\n",
       "      <td>32197736</td>\n",
       "      <td>rs436845</td>\n",
       "      <td></td>\n",
       "      <td>G</td>\n",
       "      <td>.277</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210577</th>\n",
       "      <td>6</td>\n",
       "      <td>CT</td>\n",
       "      <td>32199352</td>\n",
       "      <td>rs9267873</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>.193</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210592</th>\n",
       "      <td>6</td>\n",
       "      <td>CT</td>\n",
       "      <td>32211317</td>\n",
       "      <td>rs411326</td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>.290</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210593</th>\n",
       "      <td>6</td>\n",
       "      <td>CT</td>\n",
       "      <td>32212985</td>\n",
       "      <td>rs17576984</td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>.118</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210598</th>\n",
       "      <td>6</td>\n",
       "      <td>CT</td>\n",
       "      <td>32217092</td>\n",
       "      <td>rs6936204</td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>.237</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210602</th>\n",
       "      <td>6</td>\n",
       "      <td>CC</td>\n",
       "      <td>32220484</td>\n",
       "      <td>rs3115572</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>.397</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210604</th>\n",
       "      <td>6</td>\n",
       "      <td>CT</td>\n",
       "      <td>32223258</td>\n",
       "      <td>rs3130320</td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>.271</td>\n",
       "      <td>LOC101929163</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome genotype  position        rsid ClinVar Risk Frequency  \\\n",
       "210570          6       AG  32194854   rs3096691            G      .459   \n",
       "210573          6       AG  32197463    rs495089            G      .340   \n",
       "210574          6       AG  32197736    rs436845            G      .277   \n",
       "210577          6       CT  32199352   rs9267873            C      .193   \n",
       "210592          6       CT  32211317    rs411326            T      .290   \n",
       "210593          6       CT  32212985  rs17576984            T      .118   \n",
       "210598          6       CT  32217092   rs6936204            T      .237   \n",
       "210602          6       CC  32220484   rs3115572            C      .397   \n",
       "210604          6       CT  32223258   rs3130320            T      .271   \n",
       "\n",
       "                Gene Citations  \n",
       "210570          None         0  \n",
       "210573          None         0  \n",
       "210574          None         0  \n",
       "210577          None         0  \n",
       "210592          None         2  \n",
       "210593          None         0  \n",
       "210598          None         2  \n",
       "210602          None         1  \n",
       "210604  LOC101929163         3  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "notch\n",
    "notch_filed = notch.fillna(\"0\")\n",
    "notch_present = notch_filed\n",
    "notch_present = notch_filed[notch_filed.apply(lambda x: x.Risk in x.genotype, axis=1)]\n",
    "notch_present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('rccx.csv', 'a') as f:\n",
    "    df.to_csv(f, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
